File: core/src/main/java/com/biacode/escommons/core/repository/impl/AbstractEsRepository.java
package com.biacode.escommons.core.repository.impl;

import com.biacode.escommons.core.component.JsonComponent;
import com.biacode.escommons.core.component.SearchResponseComponent;
import com.biacode.escommons.core.model.document.AbstractEsDocument;
import com.biacode.escommons.core.model.response.DocumentsAndTotalCount;
import com.biacode.escommons.core.repository.EsRepository;
import org.elasticsearch.action.bulk.BulkRequestBuilder;
import org.elasticsearch.action.get.GetResponse;
import org.elasticsearch.action.index.IndexRequestBuilder;
import org.elasticsearch.action.search.SearchResponse;
import org.elasticsearch.client.Client;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.GenericTypeResolver;
import org.springframework.stereotype.Component;
import org.springframework.util.Assert;

import javax.annotation.Nonnull;
import java.util.*;

import static com.biacode.escommons.core.model.document.DocumentConstants.DATE_FORMAT;
import static com.biacode.escommons.core.model.document.DocumentConstants.FETCH_MAX_SIZE;
import static java.util.stream.Collectors.toMap;
import static org.elasticsearch.index.query.QueryBuilders.*;

/**
 * Created by Arthur Asatryan.
 * Date: 7/14/17
 * Time: 3:37 PM
 *
 * @param <T> the document which is subtype of AbstractEsDocument
 */
@Component
public abstract class AbstractEsRepository<T extends AbstractEsDocument> implements EsRepository<T> {

    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractEsRepository.class);

    //region Dependencies
    @Autowired
    private Client esClient;

    @Autowired
    private SearchResponseComponent searchResponseComponent;

    @Autowired
    private JsonComponent jsonComponent;
    //endregion

    //region Private properties
    private final Class<T> clazz;
    //endregion

    //region Constructors
    @SuppressWarnings("unchecked")
    public AbstractEsRepository() {
        LOGGER.debug("Initializing");
        this.clazz = (Class<T>) GenericTypeResolver.resolveTypeArgument(getClass(), AbstractEsRepository.class);
    }
    //endregion

    //region Concrete methods
    @Nonnull
    @Override
    public T save(@Nonnull final T document, @Nonnull final String indexName) {
        Assert.notNull(document, "The document should not be null");
        assertIndexNameNotNull(indexName);
        esClient.prepareIndex(indexName, getDocumentType())
                .setId(document.getUuid())
                .setSource(jsonComponent.serialize(document, clazz)).get();
        return document;
    }

    @Nonnull
    @Override
    public List<T> save(@Nonnull final List<T> documents, @Nonnull final String indexName) {
        Assert.notNull(documents, "The list of documents should not be null");
        assertIndexNameNotNull(indexName);
        final BulkRequestBuilder bulkBuilder = esClient.prepareBulk();
        documents.forEach(document -> {
            final IndexRequestBuilder indexRequestBuilder = esClient
                    .prepareIndex(indexName, getDocumentType())
                    .setId(document.getUuid())
                    .setSource(jsonComponent.serialize(document, clazz));
            bulkBuilder.add(indexRequestBuilder);
        });
        bulkBuilder.get();
        return documents;
    }

    @Nonnull
    @Override
    public String delete(@Nonnull final String id, @Nonnull final String indexName) {
        return esClient.prepareDelete(indexName, getDocumentType(), id).get().getId();
    }

    @Nonnull
    @Override
    public List<String> delete(@Nonnull final List<String> ids, @Nonnull final String indexName) {
        final BulkRequestBuilder bulkBuilder = esClient.prepareBulk();
        ids.forEach(uuid -> bulkBuilder.add(esClient.prepareDelete(indexName, getDocumentType(), uuid)));
        bulkBuilder.get();
        return ids;
    }

    @Nonnull
    @Override
    public Optional<T> findById(@Nonnull final String id, @Nonnull final String indexName) {
        assertIndexNameNotNull(indexName);
        final GetResponse response = esClient.prepareGet(indexName, getDocumentType(), id).get();
        if (!response.isExists()) {
            return Optional.empty();
        }
        return Optional.of(searchResponseComponent.convertGetResponseToDocument(response, clazz));
    }

    @Nonnull
    @Override
    public DocumentsAndTotalCount<T> findByIds(@Nonnull final List<String> ids, @Nonnull final String indexName) {
        assertIndexNameNotNull(indexName);
        final SearchResponse searchResponse = esClient.prepareSearch(indexName)
                .setQuery(boolQuery().filter(termsQuery("uuid", ids)))
                .setSize(FETCH_MAX_SIZE)
                .setTypes(getDocumentType())
                .get();
        return searchResponseComponent.convertToDocumentsAndTotalCount(searchResponse, clazz);
    }

    @Nonnull
    @Override
    public Map<Object, Object> findByField(@Nonnull final String searchField,
                                           @Nonnull final List<Object> terms,
                                           @Nonnull final String resultField,
                                           @Nonnull final String indexName,
                                           @Nonnull final String documentType) {
        final SearchResponse searchResponse = esClient.prepareSearch(indexName)
                .setTypes(documentType)
                .setQuery(boolQuery().must(matchAllQuery()).filter(termsQuery(searchField, terms)))
                .get();
        return Arrays.stream(searchResponse.getHits().getHits())
                .collect(toMap(hit -> hit.getSource().get(searchField), hit -> hit.getSource().get(resultField)));
    }
    //endregion

    //region Abstract methods
    protected abstract String getDocumentType();
    //endregion

    //region Protected methods
    protected Class<T> getClazz() {
        return clazz;
    }

    @Nonnull
    protected String[] stringValues(@Nonnull final Object... values) {
        Assert.notNull(values, "The string values should not be null");
        return Arrays.stream(values).map(this::stringValue).toArray(String[]::new);
    }

    @Nonnull
    protected String stringValue(@Nonnull final Object value) {
        Assert.notNull(value, "The string value should not be null");
        return value.toString();
    }

    @Nonnull
    protected String dateValue(@Nonnull final Object date) {
        Assert.notNull(date, "The date value should not be null");
        return date instanceof Date ? DATE_FORMAT.format(date) : date.toString();
    }

    @Nonnull
    protected String decimalValue(@Nonnull final Object decimal) {
        Assert.notNull(decimal, "The decimal value should not be null");
        return decimal.toString();
    }
    //endregion

    //region Utility methods
    private void assertIndexNameNotNull(final String indexName) {
        Assert.notNull(indexName, "The index name should not be null");
    }
    //endregion
}

File: escommons-persistence/src/main/kotlin/org/biacode/escommons/persistence/repository/impl/AbstractEsRepository.kt
package org.biacode.escommons.persistence.repository.impl

import org.biacode.escommons.core.model.document.AbstractEsDocument
import org.biacode.escommons.core.model.document.DocumentConstants
import org.biacode.escommons.persistence.repository.EsRepository
import org.biacode.escommons.toolkit.component.JsonComponent
import org.biacode.escommons.toolkit.component.SearchResponseComponent
import org.elasticsearch.action.bulk.BulkRequest
import org.elasticsearch.action.delete.DeleteRequest
import org.elasticsearch.action.get.GetRequest
import org.elasticsearch.action.get.MultiGetRequest
import org.elasticsearch.action.index.IndexRequest
import org.elasticsearch.client.RestHighLevelClient
import org.elasticsearch.common.xcontent.XContentType
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.core.GenericTypeResolver
import org.springframework.stereotype.Component
import org.springframework.util.Assert
import java.util.*

/**
 * Created by Arthur Asatryan.
 * Date: 7/14/17
 * Time: 3:37 PM
 */
@Component
abstract class AbstractEsRepository<T : AbstractEsDocument> : EsRepository<T> {

    //region Dependencies
    @Autowired
    private lateinit var esCommonsRestClient: RestHighLevelClient

    @Autowired
    private lateinit var searchResponseComponent: SearchResponseComponent

    @Autowired
    private lateinit var jsonComponent: JsonComponent
    //endregion

    //region Protected methods
    protected val clazz: Class<T>
    //endregion

    //region Constructors
    init {
        LOGGER.debug("Initializing - {}", javaClass.canonicalName)
        this.clazz = GenericTypeResolver.resolveTypeArgument(javaClass, AbstractEsRepository::class.java) as Class<T>
    }
    //endregion

    //region Abstract Methods
    fun getDocumentType(): String = "doc"
    //endregion

    //region Concrete methods
    override fun save(document: T, indexName: String): Boolean {
        val index = esCommonsRestClient
                .index(IndexRequest(indexName, DOCUMENT_TYPE, document.id).source(jsonComponent.serialize(document, clazz), XContentType.JSON))
        return index.shardInfo.failed <= 0
    }

    override fun save(documents: List<T>, indexName: String): Boolean {
        val bulkRequest = BulkRequest()
        documents.forEach {
            bulkRequest.add(
                    IndexRequest(indexName, DOCUMENT_TYPE, it.id)
                            .source(jsonComponent.serialize(it, clazz), XContentType.JSON)
            )
        }
        return !esCommonsRestClient.bulk(bulkRequest).hasFailures()
    }

    override fun delete(id: String, indexName: String): Boolean {
        return esCommonsRestClient.delete(DeleteRequest(indexName, DOCUMENT_TYPE, id)).shardInfo.failed <= 0
    }

    override fun delete(ids: List<String>, indexName: String): Boolean {
        val bulkRequest = BulkRequest()
        ids.forEach { bulkRequest.add(DeleteRequest(indexName, DOCUMENT_TYPE, it)) }
        return esCommonsRestClient.bulk(bulkRequest).hasFailures()
    }

    override fun findById(id: String, indexName: String): Optional<T> {
        val getResponse = esCommonsRestClient.get(GetRequest(indexName, DOCUMENT_TYPE, id))
        return if (!getResponse.isExists) {
            Optional.empty()
        } else Optional.of(searchResponseComponent.convertGetResponseToDocument(getResponse, clazz))
    }

    override fun findByIds(ids: List<String>, indexName: String): List<T> {
        val multiGetRequest = MultiGetRequest()
        ids.forEach { multiGetRequest.add(indexName, DOCUMENT_TYPE, it) }
        val multiGetResponse = esCommonsRestClient.multiGet(multiGetRequest)
        return multiGetResponse.responses.filter { it.response.isExists }.map { jsonComponent.deserializeFromString(it.response.sourceAsString, clazz) }
    }

    protected fun stringValues(vararg values: Any): Array<out Any> {
        Assert.notNull(values, "The string values should not be null")
        return Arrays.stream(values).map { this.stringValue(it) }.toArray()
    }

    protected fun stringValue(value: Any): String {
        Assert.notNull(value, "The string value should not be null")
        return value.toString()
    }

    protected fun dateValue(date: Any): String {
        Assert.notNull(date, "The date value should not be null")
        return if (date is Date) DocumentConstants.DATE_FORMAT.format(date) else date.toString()
    }

    protected fun dateTimeValue(date: Any): String {
        Assert.notNull(date, "The date value should not be null")
        return if (date is Date) DocumentConstants.DATE_TIME_FORMAT.format(date) else date.toString()
    }

    protected fun decimalValue(decimal: Any): String {
        Assert.notNull(decimal, "The decimal value should not be null")
        return decimal.toString()
    }

    protected fun documentType(): String = DOCUMENT_TYPE
    //endregion

    //region Companion object
    companion object {
        private val LOGGER = LoggerFactory.getLogger(AbstractEsRepository::class.java)
        private const val DOCUMENT_TYPE = "doc"
    }
    //endregion
}
